p2p/peerexchange/bookchecker.go

```diff
@@ -0,0 +1,133 @@
+package peerexchange
+
+import (
+	"context"
+	"math/rand"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+)
+
+// CheckBook periodically checks the book for dead|alive peers.
+func (d *Discovery) CheckBook(ctx context.Context) {
+	if d.cfg.CheckInterval == 0 {
+		d.logger.Info("periodically peers checking disabled")
+		return
+	}
+	ticker := time.NewTicker(d.cfg.CheckInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			d.logger.Debug("checking book")
+			d.CheckPeers(ctx)
+			d.logger.Debug("checking book done")
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// CheckPeers periodically checks the book for dead|alive peers.
+func (d *Discovery) CheckPeers(ctx context.Context) {
+	peers := d.GetRandomPeers(d.cfg.CheckPeersNumber)
+	if len(peers) == 0 {
+		d.logger.Info("no peers to check")
+		return
+	}
+	qCtx, cancel := context.WithTimeout(ctx, d.cfg.CheckTimeout)
+	defer cancel()
+	if _, err := d.crawl.query(qCtx, peers); err != nil {
+		d.logger.Error("failed to check nodes: %s", err)
+		return
+	}
+	// check peers are updated in host peerBook.
+	for peerID, addresses := range d.peersToMap(peers) {
```

Comment:
i dont understand why addresses needs to be an array. we get N random peers on L35. is it possible for that result to contain duplicates?

Reply:
can be situation, when some nodeA update local address. After updating - node will be trigger notify event and peerBook of nodeB add new address and will contain multiply addresses for nodeA. that's why peerBook contains multiply addresses per node

while our addreesBook store address in `map[peerID]*addrInfo`. in method `GetRandomPeers` L105 i check, does peerBook have extra address for nodeA. if so - add additional address for check.

than check all addresses of node. if one of them is not available - remove. if only one - update to alive address